### PR TITLE
Generative AI API

### PR DESCRIPTION
**Pull request by SidharthSawhney**
This branch supports the Generate Idea use case by directly implementing the API call. The API is Hugging Face with the 2 endpoints mistralai/Mistral-7B-Instruct-v0.1 and Salesforce/codegen-350M-mono, which are the AI models responsible for generating a business model for the given idea prompt randomly generated after clicking the "Generate Idea!" button. The business model is then displayed on the screen after which the user can regenerate another idea and corresponding business model, post the idea and business model generated or go back to the Home page. 

JavaDoc has been written for the public methods and classes and Tests for the associated entities, interactor, integration tests and view end to end tests have been written. 

Kindly let me know if anything needs to be corrected or incase there is a scope for improvement. Thank you!

### CODE REVIEW

**Comment by SidharthSawhney** [comment]
I have merged the changes in the Create Post and New window use cases and made changes to the tests and some to the code structure. This pull request is now a combination of create post and generative AI use case.

File: src/main/java/api/MistralAIAPI.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        MediaType mediaType = MediaType.parse("application/json");
+        BufferedReader bufferedReader = new BufferedReader(new FileReader("src/main/java/api/API_TOKEN"));
+        API_TOKEN = bufferedReader.readLine();
+        int i =0;
+        JSONObject generatedText = null;
+        JSONObject requestBodyJson = new JSONObject();
+
+        requestBodyJson.put("inputs", idea.getIdea() + "Write a Business model for this idea?");
+
+        while(i < 10) { //TODO: set to i < 50 during final submission
```
Comment (Tanmay-Shinde) [inline src/main/java/api/MistralAIAPI.java:33]: You might want to take care of this TODO before merging into main

File: src/main/java/api/MistralCodegenAIAPI.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        MediaType mediaType = MediaType.parse("application/json");
+        BufferedReader bufferedReader = new BufferedReader(new FileReader("src/main/java/api/API_TOKEN"));
+        API_TOKEN = bufferedReader.readLine();
+        int i = 0;
+        JSONObject generatedText = null;
+        JSONObject requestBodyJson = new JSONObject();
+
+        requestBodyJson.put("inputs", idea.getIdea() + "Write a Business model for this idea?");
+
+        while (i < 10) { //TODO: set to i < 50 during final submission
```
Comment (Tanmay-Shinde) [inline src/main/java/api/MistralCodegenAIAPI.java:45]: Same here as above

**Comment by Pranjal-R-Agrawal** [comment]
All 3 tests in GenerateIdeaViewTests are failing on my laptop. I have downloaded ideas.csv and the api key and placed them in the appropriate locatoins.

**Review by Tanmay-Shinde — CHANGES_REQUESTED** [review]
Implement the TODOs you have mentioned above (if they are still relevant and need to be changed). 

Rest all looks good to me. Javadocs and test are also well written!

File: src/main/java/api/MistralAIAPI.java (hunk 1 above)
Comment (SidharthSawhney) [reply src/main/java/api/MistralAIAPI.java:33]: setting it to 50 will slow the code for everybody as they develop, i have updated the todo comment to make the reason clear

File: src/main/java/api/MistralCodegenAIAPI.java (hunk 2 above)
Comment (SidharthSawhney) [reply src/main/java/api/MistralCodegenAIAPI.java]: setting it to 50 will slow the code for everybody as they develop, i have updated the todo comment to make the reason clear

**Review by Tanmay-Shinde — APPROVED** [review]

**Review by YathusanK — APPROVED** [review]
Use case was implemented very well and seems to function correctly. Your interactor tests and integration test also work properly and the interactor test isolates the interactor well. Javadoc was also used on all public methods or classes which is excellent. You also adhere to Clean Architecture as all dependencies point inwards to the higher levels. Your code is also well organized and follows the CA engine. Furthermore, you adhere to SOLID principles well. For example, the Single Responsibility Principle is followed since the interactor uses a class implementing the GenerativeAIAPI to generate the business model so that the interactor doesn't have multiple responsibilities. You also use design patterns like factories in your code to solve problems.

File: src/main/java/api/MistralAIAPI.java
[the remark this answers is not in the document]
Comment (SidharthSawhney) [reply src/main/java/api/MistralAIAPI.java:1]: No, it is now a legacy code

**Pull request merged**